File: Assets/src/model/enemies/Bat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Targets the nearest creature.\nHeals when it deals damage.\nGoes into Deep Sleep after 7 turns awake.", flavorText: "\"Eat, sleep, fly, repeat!\nAs far as I'm concerned, you're meat!\"\n\t - Northland nursery rhyme")]
public class Bat : AIActor, IActionPerformedHandler, IDealAttackDamageHandler {
  public Bat(Vector2Int pos) : base(pos) {
    hp = baseMaxHp = 7;
    // ClearTasks();
    faction = Faction.Enemy;
    if (MyRandom.value < 0.2f) {
      inventory.AddItem(new ItemBatTooth());
    }
  }

  int turnsUntilSleep = 7;
  public void HandleActionPerformed(BaseAction arg1, BaseAction arg2) {
    if (!(task is SleepTask)) {
      turnsUntilSleep--;
      if (turnsUntilSleep <= 0) {
        var sleep = new SleepTask(this, 5, true);
        SetTasks(sleep);
      }
    }
  }

  protected override void TaskChanged() {
    if (task is SleepTask) {
      turnsUntilSleep = 7;
    }
    base.TaskChanged();
  }

  /// bats chase and attack their closest target
  protected override ActorTask GetNextTask() {
    var target = SelectTarget();
    if (target == null) {
      return new MoveRandomlyTask(this);
    }
    if (IsNextTo(target)) {
      return new AttackTask(this, target);
    }
    // chase until you are next to any target
    return new ChaseDynamicTargetTask(this, SelectTarget);
  }

  Actor SelectTarget() {
    var potentialTargets = new HashSet<Actor>(floor
      .ActorsInCircle(pos, 7)
      .Where((t) => floor.TestVisibility(pos, t.pos) && !(t is Bat)));

    if (potentialTargets.Contains(GameModel.main.player) && GameModel.main.player.isCamouflaged) {
      potentialTargets.Remove(GameModel.main.player);
    }

    if (potentialTargets.Any()) {
      return potentialTargets.Aggregate((t1, t2) => DistanceTo(t1) < DistanceTo(t2) ? t1 : t2);
    }
    return null;
  }

  public void HandleDealAttackDamage(int dmg, Body target) {
    if (target is Actor && dmg > 0 && target != this) {
      Heal(1);
    }
  }

  internal override (int, int) BaseAttackDamage() {
    return (1, 1);
  }
}

[Serializable]
[ObjectInfo("bat-tooth", "Sharp with a little hole on the end to extract blood.")]
internal class ItemBatTooth : EquippableItem, IWeapon, IDealAttackDamageHandler {
  internal override string GetStats() => "Heals HP equal to damage dealt.\nConsumed on use.";
  public ItemBatTooth() {
    durability = maxDurability;
  }

  public void HandleDealAttackDamage(int dmg, Body target) {
    if (target is Actor && dmg > 0) {
      GameModel.main.player.Heal(dmg);
    }
    Destroy();
  }

  public (int, int) AttackSpread => (1, 1);

  public int durability { get; set; }

  public int maxDurability => 1;
  public override EquipmentSlot slot => EquipmentSlot.Weapon;
}

File: web/src/model/enemies/Bat.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { ChaseDynamicTargetTask } from '../tasks/ChaseDynamicTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { SleepTask } from '../tasks/SleepTask';
import { ACTION_PERFORMED_HANDLER, DEAL_ATTACK_DAMAGE_HANDLER, type IActionPerformedHandler, type IDealAttackDamageHandler } from '../Actor';
import { CollisionLayer } from '../../core/types';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';
import type { BaseAction } from '../BaseAction';
import type { Body } from '../Body';
import type { Actor } from '../Actor';

/**
 * Flying enemy. Targets nearest creature. Heals on damage dealt.
 * Goes into deep sleep after 5 turns awake.
 * Port of C# Bat.cs.
 */
export class Bat extends AIActor implements IActionPerformedHandler, IDealAttackDamageHandler {
  readonly [ACTION_PERFORMED_HANDLER] = true as const;
  readonly [DEAL_ATTACK_DAMAGE_HANDLER] = true as const;

  get baseMovementLayer(): CollisionLayer {
    return CollisionLayer.Flying;
  }

  private turnsUntilSleep = 5;

  constructor(pos: Vector2Int) {
    super(pos);
    this.hp = this._baseMaxHp = 3;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  handleActionPerformed(_final: BaseAction, _initial: BaseAction): void {
    if (!(this.task instanceof SleepTask)) {
      this.turnsUntilSleep--;
      if (this.turnsUntilSleep <= 0) {
        this.setTasks(new SleepTask(this, 5, true));
      }
    }
  }

  protected taskChanged(): void {
    if (this.task instanceof SleepTask) {
      this.turnsUntilSleep = 5;
    }
    super.taskChanged();
  }

  handleDealAttackDamage(dmg: number, target: Body): void {
    if (target !== this && (target as any).faction !== undefined && dmg > 0) {
      this.heal(1);
    }
  }

  protected getNextTask(): ActorTask {
    const target = this.selectTarget();
    if (!target) {
      return new MoveRandomlyTask(this);
    }
    if (this.isNextTo(target)) {
      return new AttackTask(this, target);
    }
    return new ChaseDynamicTargetTask(this, () => this.selectTarget()!);
  }

  private selectTarget(): Actor | null {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      return player;
    }
    return null;
  }
}

entityRegistry.register('Bat', Bat);
